### PR TITLE
Base class for generated attribute records

### LINKED ISSUE #685: Generated code looses attributes that are not present when running the nd-codgen tool.
## The problem

The generated code has a class with properties that match the attributes for tje entities in each domain. Some entities will not always have all attributes present. eg a light entity only has a brightness attribute when it is turned on. As a result when re-generating the code while all lights are off the generated LightEntityEttributes class will not have a Brightness property. Any used code that uses this property will then fail to compile.
## Environment

22.6.3

## Traceback/Error logs

```txt

```

### PR DESCRIPTION
**Pull request by FrankBakkerNl**
## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #685
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code compiles without warnings (code quality chek)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/net-daemon/docs

### CODE REVIEW

File: src/HassModel/NetDeamon.HassModel/Entities/Core/LightAttributesBase.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
+﻿namespace NetDaemon.HassModel.Entities.Core;
+#pragma warning disable CS1591 
+
+public record LightAttributesBase
```
Comment (helto4real) [inline src/HassModel/NetDeamon.HassModel/Entities/Core/LightAttributesBase.cs:4]: I could not find any more attributes on my setup. This should be fine.

File: src/HassModel/NetDaemon.HassModel.Tests/CodeGenerator/CodeGeneratorTest.cs [hunk 2]
```diff
@@ -207,6 +209,52 @@ public void Run(IHaContext ha)
         AssertCodeCompiles(generatedCode, appCode);
     }
 
+    
+    [Fact]
+    public void TestAttributeClassGenerationSkipBaseProperties()
```
Comment (helto4real) [inline src/HassModel/NetDaemon.HassModel.Tests/CodeGenerator/CodeGeneratorTest.cs:214]: Should we also add an test to test if we do not UseAttributeBaseClasses == true? To make sure the old stuff works?

**Review by helto4real — COMMENTED** [review]
Looks good. Have few comments

File: src/HassModel/NetDaemon.HassModel.CodeGenerator/_appsettings.Development.json [hunk 3]
```diff
@@ -8,6 +8,7 @@
     },
     "CodeGeneration": {
         "Namespace": "HomeAssistantGenerated",
-        "OutputFile": "HomeAssistantGenerated.cs"
+        "OutputFile": "HomeAssistantGenerated.cs",
+        "UseAttributeBaseClasses" : "true"
```
[the remark this answers is not in the document]
Comment (FrankBakkerNl) [reply src/HassModel/NetDaemon.HassModel.CodeGenerator/_appsettings.Development.json:12]: That should be done in the template right?

File: src/HassModel/NetDaemon.HassModel.CodeGenerator/EntitiesGenerator.cs [hunk 4]
```diff
@@ -1,13 +1,15 @@
-﻿using NetDaemon.Client.HomeAssistant.Model;
+﻿using Microsoft.CodeAnalysis.CSharp;
+using NetDaemon.Client.HomeAssistant.Model;
+using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;
```
[the remark this answers is not in the document]
Comment (FrankBakkerNl) [reply src/HassModel/NetDaemon.HassModel.CodeGenerator/EntitiesGenerator.cs:3]: I think you mean MS first and then ND?

File: src/HassModel/NetDaemon.HassModel.Tests/CodeGenerator/CodeGeneratorTest.cs (hunk 2 above)
Comment (FrankBakkerNl) [reply src/HassModel/NetDaemon.HassModel.Tests/CodeGenerator/CodeGeneratorTest.cs:214]: That is covered (more or less) by an existing test, I will rename it to make that explicit

File: src/HassModel/NetDaemon.HassModel.CodeGenerator/Program.cs [hunk 5]
```diff
@@ -50,7 +50,7 @@ IConfigurationRoot GetConfigurationRoot()
 {
```
[the remark this answers is not in the document]
Comment (FrankBakkerNl) [reply src/HassModel/NetDaemon.HassModel.CodeGenerator/Program.cs:50]: Maybe we should do that when we decide is we want to change the default to true

File: src/HassModel/NetDaemon.HassModel.CodeGenerator/_appsettings.Development.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (helto4real) [reply src/HassModel/NetDaemon.HassModel.CodeGenerator/_appsettings.Development.json:12]: Yes I will fix that.

File: src/HassModel/NetDaemon.HassModel.CodeGenerator/Program.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (helto4real) [reply src/HassModel/NetDaemon.HassModel.CodeGenerator/Program.cs:50]: Yea we could.

**Review by helto4real — APPROVED** [review]
Looks good! I will make PR to the template to include the setting but setting it to default false

**Comment by phillipsawyer** [comment]
This is fantastic!

The only other really important one I can think of would be MediaPlayers which, for example, do not have a VolumeLevel if they are off.

I am constantly having to play something on a media player before I run nd-codegen.

**Pull request merged**